### PR TITLE
Update crate_universe to determine dep+feature trees per host

### PR DESCRIPTION
**Pull request by UebelAndre**
This change expands the use of `cargo tree` to resolve features for different combinations of `host -> target` platform triples where before `cargo-bazel` was only capable of resolving host dependencies for the current host platform and not a foreign platform.

A lengthy description of how this was done can be found at `crate_universe/src/metadata/cargo_tree_resolver.rs - TreeResolver::create_rustc_wrapper`.

closes https://github.com/bazelbuild/rules_rust/issues/2212

### CODE REVIEW

File: crate_universe/src/metadata/cargo_tree_resolver.rs [hunk 1]
```diff
@@ -66,71 +107,195 @@ impl TreeResolver {
...
+    /// intercept this command and replace the `host:` string with `host: ${HOST_TRIPLE}` where `HOST_TRIPLE`
+    /// is an environment variable set by this class.
+    ///
+    /// [@cargo//src/cargo/core/compiler/build_context/target_info.rs#L168-L218](https://github.com/rust-lang/cargo/blob/0.81.0/src/cargo/core/compiler/build_context/target_info.rs#L168-L218)
+    /// ```ignore
+    /// rust - --crate-name ___ --print=file-names
+    /// ```
+    ///
+    /// As seen in the associated `target_info.rs` link, this command is generally a lot more complicated but
+    /// this prefix is what's used to identify the a command to intercept. Cargo will make this call multiple times
```
Comment (illicitonion) [inline crate_universe/src/metadata/cargo_tree_resolver.rs:284]:
```suggestion
    /// this prefix is what's used to identify the command to intercept. Cargo will make this call multiple times
```

File: crate_universe/src/metadata/cargo_tree_resolver.rs [hunk 2]
```diff
@@ -66,71 +107,195 @@ impl TreeResolver {
...
+        use std::os::unix::fs::PermissionsExt;
+
+        let perms = std::fs::Permissions::from_mode(0o700);
+        std::fs::set_permissions(&wrapper, perms)
+            .context("Failed to modify permissions of rustc wrapper")?;
+
+        Ok(wrapper)
+    }
+
+    /// Create a wrapper for `rustc` which can intercept commands used to identify the host
```
Comment (illicitonion) [inline crate_universe/src/metadata/cargo_tree_resolver.rs:253]: Can you link to https://github.com/rust-lang/cargo/issues/14527?

File: crate_universe/src/metadata/cargo_tree_resolver.rs [hunk 3]
```diff
@@ -503,18 +697,74 @@ impl Ord for DependencyDetailWithOrd {
...
+    /// Collecting dependencies for the target platform.
+    Target(CrateId),
+
+    /// Collecting dependencies for the host platform (e.g. `[build-dependency]` and `proc-macro`).
+    Host(CrateId),
+
+    /// Similar to [TreeDepCompileKind::Target] represents an edge where
+    /// dependencies are to be collected for host. E.g. a crate which has
+    /// a `[build-dependency] or `(proc-macro)` dependency.
+    TargetWithHostDep(CrateId),
```
Comment (illicitonion) [inline crate_universe/src/metadata/cargo_tree_resolver.rs:713]: I'm not quite clear on what these deps are, the docs read that they're the same as `Host`? Can you try clarifying the docs?

**Review by illicitonion — COMMENTED** [review]
Fancy! Exciting! Only slightly terrifying!

Just one docs question really :)

File: crate_universe/src/metadata/cargo_tree_resolver.rs (hunk 2 above)
Comment (UebelAndre) [reply crate_universe/src/metadata/cargo_tree_resolver.rs:253]: Done!

File: crate_universe/src/metadata/cargo_tree_resolver.rs (hunk 3 above)
Comment (UebelAndre) [reply crate_universe/src/metadata/cargo_tree_resolver.rs]: Added some more words 😄

**Review by illicitonion — APPROVED** [review]
Looks great! Thanks!

**Pull request merged**